pkger/service.go

```diff
@@ -1859,44 +1895,111 @@ func (s *Service) applyNotificationRules(rules []*stateRule) applier {
 		return nil
 	}
 
+	rollbackFn := func(_ influxdb.ID) error {
+		return s.rollbackNotificationRules(ctx, userID, rollbackEndpoints)
+	}
+
 	return applier{
 		creater: creater{
 			entries: len(rules),
 			fn:      createFn,
 		},
 		rollbacker: rollbacker{
-			resource: resource,
-			fn: func(_ influxdb.ID) error {
-				return s.rollbackNotificationRules(rollbackEndpoints)
-			},
+			fn: func(_ influxdb.ID) error { return nil },
 		},
-	}
+	}, rollbackFn
 }
 
 func (s *Service) applyNotificationRule(ctx context.Context, r *stateRule, userID influxdb.ID) (influxdb.NotificationRule, error) {
-	influxRule := influxdb.NotificationRuleCreate{
-		NotificationRule: r.toInfluxRule(),
-		Status:           r.parserRule.Status(),
-	}
-	err := s.ruleSVC.CreateNotificationRule(ctx, influxRule, userID)
-	if err != nil {
-		return nil, err
+	switch r.stateStatus {
+	case StateStatusRemove:
+		if err := s.ruleSVC.DeleteNotificationRule(ctx, r.ID()); err != nil {
+			return nil, ierrors.Wrap(err, "failed to remove notification rule")
+		}
+		return r.existing, nil
+	case StateStatusExists:
+		ruleCreate := influxdb.NotificationRuleCreate{
+			NotificationRule: r.toInfluxRule(),
+			Status:           r.parserRule.Status(),
+		}
+		influxRule, err := s.ruleSVC.UpdateNotificationRule(ctx, r.ID(), ruleCreate, userID)
+		if err != nil {
+			return nil, ierrors.Wrap(err, "failed to update notification rule")
+		}
+		return influxRule, nil
+	default:
+		influxRule := influxdb.NotificationRuleCreate{
+			NotificationRule: r.toInfluxRule(),
+			Status:           r.parserRule.Status(),
+		}
+		err := s.ruleSVC.CreateNotificationRule(ctx, influxRule, userID)
+		if err != nil {
+			return nil, err
+		}
+		return influxRule.NotificationRule, nil
 	}
-
-	return influxRule, nil
 }
 
-func (s *Service) rollbackNotificationRules(rules []*stateRule) error {
+func (s *Service) rollbackNotificationRules(ctx context.Context, userID influxdb.ID, rules []*stateRule) error {
+	rollbackFn := func(r *stateRule) error {
+		existingRuleFn := func(endpointID influxdb.ID) influxdb.NotificationRule {
```

Comment:
this is rather unfortunate. We have a getter for the rule's endpointID but no setter... this works around that limiation